File: frontend/src/api/bom.ts
import { client } from './client'

export interface BomHeader {
  id: string; productVariantId: string; variantCode: string; variantName: string
  currentVersionId: string | null; currentVersionStatus: string | null; currentVersionNumber: number | null
  isActive: boolean; notes: string | null; createdAt: string
}

export interface BomLine {
  id: string; lineNumber: number; itemId: string; itemCode: string; itemName: string
  quantity: number; unitId: string; unitAbbreviation: string; wastePercent: number; notes: string | null
}

export interface BomVersion {
  id: string; bomHeaderId: string; versionNumber: number; status: string
  effectiveFrom: string; effectiveTo: string | null; changeReason: string | null
  createdAt: string; updatedAt: string; lines: BomLine[]
}

export interface BomVersionSummary {
  id: string; versionNumber: number; status: string
  effectiveFrom: string; effectiveTo: string | null; changeReason: string | null
  createdAt: string; lineCount: number
}

export const bomApi = {
  getHeaders: (productVariantId?: string) =>
    client.get<BomHeader[]>('/boms', { params: { productVariantId } }).then(r => r.data),

  getHeader: (id: string) => client.get<BomHeader>(`/boms/${id}`).then(r => r.data),

  createHeader: (req: { productVariantId: string; notes?: string }) =>
    client.post<BomHeader>('/boms', req).then(r => r.data),

  getVersions: (bomId: string) =>
    client.get<BomVersionSummary[]>(`/boms/${bomId}/versions`).then(r => r.data),

  getVersion: (bomId: string, versionId: string) =>
    client.get<BomVersion>(`/boms/${bomId}/versions/${versionId}`).then(r => r.data),

  createVersion: (bomId: string, req: { effectiveFrom: string; changeReason?: string }) =>
    client.post<BomVersion>(`/boms/${bomId}/versions`, req).then(r => r.data),

  activateVersion: (bomId: string, versionId: string, req: { changeReason?: string }) =>
    client.post<BomVersion>(`/boms/${bomId}/versions/${versionId}/activate`, req).then(r => r.data),

  addLine: (bomId: string, versionId: string, req: { itemId: string; quantity: number; unitId: string; wastePercent: number; notes?: string }) =>
    client.post<BomLine>(`/boms/${bomId}/versions/${versionId}/lines`, req).then(r => r.data),

  updateLine: (bomId: string, versionId: string, lineId: string, req: { quantity: number; unitId: string; wastePercent: number; notes?: string }) =>
    client.put<BomLine>(`/boms/${bomId}/versions/${versionId}/lines/${lineId}`, req).then(r => r.data),

  deleteLine: (bomId: string, versionId: string, lineId: string) =>
    client.delete(`/boms/${bomId}/versions/${versionId}/lines/${lineId}`),
}

File: backend/src/Csw.Api/Domain/Entities/Bom.cs
namespace Csw.Api.Domain.Entities;

public enum BomVersionStatus { Draft, Active, Superseded }

public class BomHeader
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid ProductVariantId { get; set; }
    public ProductVariant ProductVariant { get; set; } = null!;
    public Guid? CurrentVersionId { get; set; }
    public BomVersion? CurrentVersion { get; set; }
    public bool IsActive { get; set; } = true;
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    public List<BomVersion> Versions { get; set; } = [];
}

public class BomVersion
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid BomHeaderId { get; set; }
    public BomHeader BomHeader { get; set; } = null!;
    public int VersionNumber { get; set; }
    public BomVersionStatus Status { get; set; } = BomVersionStatus.Draft;
    public DateOnly EffectiveFrom { get; set; } = DateOnly.FromDateTime(DateTime.UtcNow);
    public DateOnly? EffectiveTo { get; set; }
    public Guid? CreatedByUserId { get; set; }
    public Guid? ApprovedByUserId { get; set; }
    public DateTime? ApprovedAt { get; set; }
    public string? ChangeReason { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    public List<BomLine> Lines { get; set; } = [];
}

public class BomLine
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid BomVersionId { get; set; }
    public BomVersion BomVersion { get; set; } = null!;
    public int LineNumber { get; set; }
    public Guid ItemId { get; set; }
    public Item Item { get; set; } = null!;
    public decimal Quantity { get; set; }
    public Guid UnitId { get; set; }
    public Unit Unit { get; set; } = null!;
    public decimal WastePercent { get; set; } = 0;
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}
